Fix race condtion when mutliple local endpoints are created

aswinsuryan opened the pull request:
There is a race condition when the worker nodes get restarted one by one. More than one submariner endpoint is created in one cluster as two gateways tries to be the leader. One yields later and this triggers a endpoint delete event. This causes routes to be cleared in both local and remote cluster. Later an endpoint update even comes for the active gateway, and we are using the update event in the fix

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

aswinsuryan committed:
Fix race condtion when mutliple local endpoints are created

There is a race condition when the worker nodes get restarted one by one.
More than one submariner endpoint is created in one cluster as two gateways
tries to be the leader. One yields later and this triggers a endpoint delete
event. This causes routes to be cleared in both local and remote cluster.
Later an endpoint update even comes for the active gateway, and we are using
the update event in the fix

Signed-off-by: Aswin Suryanarayanan <asuryana@redhat.com>

tpantelis commented on pkg/routeagent_driver/handlers/kubeproxy/endpoint_handler.go:
```diff
@@ -71,7 +72,48 @@ func (kp *SyncHandler) LocalEndpointCreated(endpoint *submV1.Endpoint) error {
 
-func (kp *SyncHandler) LocalEndpointUpdated(_ *submV1.Endpoint) error {
+func (kp *SyncHandler) LocalEndpointUpdated(endpoint *submV1.Endpoint) error {
+	if kp.localEndpoint == nil || kp.localEndpoint.Name != endpoint.Name {
```
To be consistent, we should protect access to `localEndpoint` with the `syncHandlerMutex`.

tpantelis commented on pkg/routeagent_driver/handlers/kubeproxy/endpoint_handler.go:
```diff
@@ -30,6 +30,7 @@ func (kp *SyncHandler) LocalEndpointCreated(endpoint *submV1.Endpoint) error {
 	kp.syncHandlerMutex.Lock()
 	defer kp.syncHandlerMutex.Unlock()
 	kp.localCableDriver = endpoint.Spec.Backend
+	kp.localEndpoint = endpoint
```
We don't need to cache the `localCableDriver` now that the whole endpoint is cached.

tpantelis commented:
> More than one submariner endpoint is created in one cluster as two gateways tries to be the leader. One yields later and this triggers a endpoint delete event.

Can you explain this in more detail? Have you observed a scenario where 2 gateways are simultaneously the leader?

Whatever the exact scenario/sequence, I wonder if https://github.com/submariner-io/submariner/pull/2763 might alleviate it.

aswinsuryan commented:
> > More than one submariner endpoint is created in one cluster as two gateways tries to be the leader. One yields later and this triggers a endpoint delete event.
> 
> Can you explain this in more detail? Have you observed a scenario where 2 gateways are simultaneously the leader?
> 
> Whatever the exact scenario/sequence, I wonder if #2763 might alleviate it.

When nodes are restarted, there are disruption in api server connection and during this time,  two leaders are present simultaneously and two local endpoints are created. Later one of them yields as it is not able to get the gateway lock.

I think https://github.com/submariner-io/submariner/pull/2763 will not fix it. @sridhargaddam  any thoughts?

tpantelis commented:
> When nodes are restarted, there are disruption in api server connection and during this time, two leaders are present simultaneously and two local endpoints are created. Later one of them yields as it is not able to get the gateway lock.
> 

Are you saying that there are 2 local endpoints present in the cluster at the same time? When the gateway acquires the leader lock, it first deletes any other endpoints before creating its own. Also it doesn't delete its endpoint if it loses the leader lock and/or exits. I'm trying to see a scenario where 2 local endpoints coexist simultaneously. The whole point of using leader election is to prevent 2 gateway controllers from running critical code simultaneously. If there's a hole in this process then I think we need to understand this. I'm wary of just trying to "band-aid" it in the route-agent.

If the API server isn't available then it shouldn't be able to acquire the leader lock nor create endpoints as those operations require access to the API server.

What could happen is if the first gateway acquires the lock and creates an endpoint then later loses connection to the API server and can't renew the lock after 5 sec (`RenewDeadline`) and exits. Then the second gateway acquires the lock after 10 sec (`LeaseDuration`) and creates another endpoint. But this is a fail-over scenario, ie there's aren't 2 leaders present simultaneously.

Can you provide a detailed step-by-step sequence of the scenario and any log messages?

aswinsuryan commented:
> > When nodes are restarted, there are disruption in api server connection and during this time, two leaders are present simultaneously and two local endpoints are created. Later one of them yields as it is not able to get the gateway lock.
> 
> Are you saying that there are 2 local endpoints present in the cluster at the same time? When the gateway acquires the leader lock, it first deletes any other endpoints before creating its own. Also it doesn't delete its endpoint if it loses the leader lock and/or exits. I'm trying to see a scenario where 2 local endpoints coexist simultaneously. The whole point of using leader election is to prevent 2 gateway controllers from running critical code simultaneously. If there's a hole in this process then I think we need to understand this. I'm wary of just trying to "band-aid" it in the route-agent.
> 
> If the API server isn't available then it shouldn't be able to acquire the leader lock nor create endpoints as those operations require access to the API server.
> 
> What could happen is if the first gateway acquires the lock and creates an endpoint then later loses connection to the API server and can't renew the lock after 5 sec (`RenewDeadline`) and exits. Then the second gateway acquires the lock after 10 sec (`LeaseDuration`) and creates another endpoint. But this is a fail-over scenario, ie there's aren't 2 leaders present simultaneously.
> 
> Can you provide a detailed step-by-step sequence of the scenario and any log messages?

This is the log from node where gateway got elected and later yielded 
`[90m2023-10-17T10:23:08.621Z[0m [32mINF[0m ..-source/app/main.go:429 main                 Using namespace "submariner-operator" for the leader election lock
I1017 10:23:08.621704       1 leaderelection.go:245] attempting to acquire leader lease submariner-operator/submariner-gateway-lock...
I1017 10:23:08.731886       1 leaderelection.go:255] successfully acquired lease submariner-operator/submariner-gateway-lock
[90m2023-10-17T10:23:08.731Z[0m [33mDBG[0m ..ols/record/event.go:298 main                 Event(v1.ObjectReference{Kind:"Lease", Namespace:"submariner-operator", Name:"submariner-gateway-lock", UID:"25695380-3b5e-4125-8a16-e5ea46cd37cc", APIVersion:"coordination.k8s.io/v1", ResourceVersion:"5520789", FieldPath:""}): type: 'Normal' reason: 'LeaderElection' compute-1-submariner-gateway became leader
[90m2023-10-17T10:23:08.732Z[0m [32mINF[0m ..reswan/libreswan.go:127 libreswan            Using NATT UDP port 4500
[90m2023-10-17T10:23:08.732Z[0m [32mINF[0m ..gine/cableengine.go:104 CableEngine          CableEngine controller started, driver: "libreswan"
[90m2023-10-17T10:23:08.732Z[0m [32mINF[0m ..public_ip_watcher.go:62 Endpoint             Starting the public IP watcher.
[90m2023-10-17T10:23:08.732Z[0m [32mINF[0m ..r/datastoresyncer.go:73 DSSyncer             Starting the datastore syncer
[90m2023-10-17T10:23:08.733Z[0m [32mINF[0m ..ers/tunnel/tunnel.go:40 Tunnel               Starting the tunnel controller
[90m2023-10-17T10:23:08.834Z[0m [32mINF[0m ..er/healthchecker.go:110 HealthChecker        CableEngine HealthChecker started with PingInterval: 1, MaxPacketLossCount: 5
[90m2023-10-17T10:23:08.835Z[0m [32mINF[0m ..thchecker/pinger.go:106 HealthChecker        Starting pinger for IP "242.1.255.254"
[90m2023-10-17T10:23:08.835Z[0m [32mINF[0m ..er/healthchecker.go:159 HealthChecker        CableEngine HealthChecker started pinger for CableName: "submariner-cable-dnd-prsurve-dev-2-10-1-114-105" with HealthCheckIP "242.1.255.254"
[90m2023-10-17T10:23:08.863Z[0m [32mINF[0m ..ery/natdiscovery.go:162 NAT                  Starting NAT discovery for endpoint "submariner-cable-dnd-prsurve-dev-2-10-1-114-105"
[90m2023-10-17T10:23:09.008Z[0m [32mINF[0m .._update_federator.go:72 Federator            local -> broker: Updated Cluster "submariner-broker/dnd-prsurve-dev-1" 
[90m2023-10-17T10:23:09.032Z[0m [32mINF[0m .._update_federator.go:72 Federator            broker -> local: Updated Cluster "submariner-operator/dnd-prsurve-dev-2" 
[90m2023-10-17T10:23:09.061Z[0m [32mINF[0m ../datastoresyncer.go:241 DSSyncer             Ensuring we are the only endpoint active for this cluster
[90m2023-10-17T10:23:09.062Z[0m [32mINF[0m ../datastoresyncer.go:316 DSSyncer             Creating local submariner Cluster: types.SubmarinerCluster{ID:"dnd-prsurve-dev-1", Spec:v1.ClusterSpec{ClusterID:"dnd-prsurve-dev-1", ColorCodes:[]string{"blue"}, ServiceCIDR:[]string{"172.30.0.0/16"}, ClusterCIDR:[]string{"10.128.0.0/14"}, GlobalCIDR:[]string{"242.0.0.0/16"}}} 
[90m2023-10-17T10:23:09.095Z[0m [32mINF[0m .._update_federator.go:72 Federator            broker -> local: Updated Cluster "submariner-operator/dnd-prsurve-dev-1" 
[90m2023-10-17T10:23:09.095Z[0m [32mINF[0m ../datastoresyncer.go:329 DSSyncer             Creating local submariner Endpoint: types.SubmarinerEndpoint{Spec:v1.EndpointSpec{ClusterID:"dnd-prsurve-dev-1", CableName:"submariner-cable-dnd-prsurve-dev-1-10-1-114-39", HealthCheckIP:"", Hostname:"compute-1", Subnets:[]string{"242.0.0.0/16"}, PrivateIP:"10.1.114.39", PublicIP:"66.187.232.130", NATEnabled:false, Backend:"libreswan", BackendConfig:map[string]string{"natt-discovery-port":"4490", "preferred-server":"false", "udp-port":"4500"}}} 
[90m2023-10-17T10:23:09.103Z[0m [32mINF[0m .._update_federator.go:72 Federator            local -> broker: Updated Endpoint "submariner-broker/dnd-prsurve-dev-1-submariner-cable-dnd-prsurve-dev-1-10-1-114-39" 
[90m2023-10-17T10:23:09.143Z[0m [32mINF[0m .._update_federator.go:72 Federator            broker -> local: Updated Endpoint "submariner-operator/dnd-prsurve-dev-1-submariner-cable-dnd-prsurve-dev-1-10-1-114-39" 
[90m2023-10-17T10:23:09.158Z[0m [32mINF[0m .._update_federator.go:72 Federator            broker -> local: Updated Endpoint "submariner-operator/dnd-prsurve-dev-2-submariner-cable-dnd-prsurve-dev-2-10-1-114-105" 
[90m2023-10-17T10:23:09.217Z[0m [32mINF[0m .._update_federator.go:72 Federator            local -> broker: Updated Endpoint "submariner-broker/dnd-prsurve-dev-1-submariner-cable-dnd-prsurve-dev-1-10-1-114-39" 
[90m2023-10-17T10:23:09.285Z[0m [32mINF[0m ../datastoresyncer.go:105 DSSyncer             Datastore syncer started
[90m2023-10-17T10:23:09.288Z[0m [32mINF[0m ..ncer/node_handler.go:73 DSSyncer             Updating the endpoint HealthCheckIP to globalIP "242.0.255.254"
[90m2023-10-17T10:23:09.288Z[0m [32mINF[0m ../datastoresyncer.go:329 DSSyncer             Creating local submariner Endpoint: types.SubmarinerEndpoint{Spec:v1.EndpointSpec{ClusterID:"dnd-prsurve-dev-1", CableName:"submariner-cable-dnd-prsurve-dev-1-10-1-114-39", HealthCheckIP:"242.0.255.254", Hostname:"compute-1", Subnets:[]string{"242.0.0.0/16"}, PrivateIP:"10.1.114.39", PublicIP:"66.187.232.130", NATEnabled:false, Backend:"libreswan", BackendConfig:map[string]string{"natt-discovery-port":"4490", "preferred-server":"false", "udp-port":"4500"}}} 
[90m2023-10-17T10:23:09.302Z[0m [33mDBG[0m ..ry/request_handle.go:52 NAT                  Received request from 10.1.114.105:4490 - REQUEST_NUMBER: 0x3f5a538c62273a99, SENDER: "submariner-cable-dnd-prsurve-dev-2-10-1-114-105", RECEIVER: "submariner-cable-dnd-prsurve-dev-1-10-1-114-39"
[90m2023-10-17T10:23:09.303Z[0m [33mDBG[0m ..y/request_handle.go:117 NAT                  Sending response to 10.1.114.105:4490 - REQUEST_NUMBER: 0x3f5a538c62273a99, RESPONSE: OK, SENDER: "submariner-cable-dnd-prsurve-dev-1-10-1-114-39", RECEIVER: "submariner-cable-dnd-prsurve-dev-2-10-1-114-105"
[90m2023-10-17T10:23:09.323Z[0m [32mINF[0m .._update_federator.go:72 Federator            broker -> local: Updated Endpoint "submariner-operator/dnd-prsurve-dev-1-submariner-cable-dnd-prsurve-dev-1-10-1-114-39" 
[90m2023-10-17T10:23:09.375Z[0m [32mINF[0m .._update_federator.go:72 Federator            local -> broker: Updated Endpoint "submariner-broker/dnd-prsurve-dev-1-submariner-cable-dnd-prsurve-dev-1-10-1-114-39" 
[90m2023-10-17T10:23:09.523Z[0m [33mDBG[0m ..ery/request_send.go:115 NAT                  Sending request - REQUEST_NUMBER: 0xc089e199fd91e6c6, SENDER: "submariner-cable-dnd-prsurve-dev-1-10-1-114-39", RECEIVER: "submariner-cable-dnd-prsurve-dev-2-10-1-114-105", USING_SRC: 10.1.114.39:4490, USING_DST: 10.1.114.105:4490
[90m2023-10-17T10:23:09.523Z[0m [33mDBG[0m ..ery/request_send.go:115 NAT                  Sending request - REQUEST_NUMBER: 0xc089e199fd91e6c7, SENDER: "submariner-cable-dnd-prsurve-dev-1-10-1-114-39", RECEIVER: "submariner-cable-dnd-prsurve-dev-2-10-1-114-105", USING_SRC: 10.1.114.39:4490, USING_DST: 66.187.232.130:4490
[90m2023-10-17T10:23:09.548Z[0m [33mDBG[0m ..y/response_handle.go:31 NAT                  Received response from 10.1.114.105:4490 - REQUEST_NUMBER: 0xc089e199fd91e6c6, RESPONSE: OK, SENDER: "submariner-cable-dnd-prsurve-dev-2-10-1-114-105", RECEIVER: "submariner-cable-dnd-prsurve-dev-1-10-1-114-39"
[90m2023-10-17T10:23:09.548Z[0m [33mDBG[0m ../remote_endpoint.go:184 NAT                  selected private IP "10.1.114.105" for endpoint "submariner-cable-dnd-prsurve-dev-2-10-1-114-105"
[90m2023-10-17T10:23:09.548Z[0m [32mINF[0m ..gine/cableengine.go:195 CableEngine          Installing Endpoint cable "submariner-cable-dnd-prsurve-dev-2-10-1-114-105"
[90m2023-10-17T10:23:09.548Z[0m [32mINF[0m ..reswan/libreswan.go:579 libreswan            Starting Pluto
002 listening for IKE messages
002 forgetting secrets
002 loading secrets from "/etc/ipsec.secrets"
002 loading secrets from "/etc/ipsec.d/submariner.secrets"
[90m2023-10-17T10:23:10.159Z[0m [32mINF[0m ..reswan/libreswan.go:371 libreswan            Creating connection(s) for {"metadata":{"name":"dnd-prsurve-dev-2-submariner-cable-dnd-prsurve-dev-2-10-1-114-105","namespace":"submariner-operator","uid":"ed71ddf8-3c47-4d21-ac4f-e94d3d0245b1","resourceVersion":"98223","generation":2,"creationTimestamp":"2023-10-13T11:14:34Z","labels":{"submariner-io/clusterID":"dnd-prsurve-dev-2"},"managedFields":[{"manager":"submariner-gateway","operation":"Update","apiVersion":"submariner.io/v1","time":"2023-10-13T11:14:35Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:labels":{".":{},"f:submariner-io/clusterID":{}}},"f:spec":{".":{},"f:backend":{},"f:backend_config":{".":{},"f:natt-discovery-port":{},"f:preferred-server":{},"f:udp-port":{}},"f:cable_name":{},"f:cluster_id":{},"f:healthCheckIP":{},"f:hostname":{},"f:nat_enabled":{},"f:private_ip":{},"f:public_ip":{},"f:subnets":{}}}}]},"spec":{"cluster_id":"dnd-prsurve-dev-2","cable_name":"submariner-cable-dnd-prsurve-dev-2-10-1-114-105","healthCheckIP":"242.1.255.254","hostname":"compute-0","subnets":["242.1.0.0/16"],"private_ip":"10.1.114.105","public_ip":"66.187.232.130","nat_enabled":false,"backend":"libreswan","backend_config":{"natt-discovery-port":"4490","preferred-server":"false","udp-port":"4500"}}} in bi-directional mode
[90m2023-10-17T10:23:10.159Z[0m [32mINF[0m ..reswan/libreswan.go:435 libreswan            Executing whack with args: [--psk --encrypt --name submariner-cable-dnd-prsurve-dev-2-10-1-114-105-0-0 --id 10.1.114.39 --host 10.1.114.39 --client 242.0.0.0/16 --ikeport 4500 --to --id 10.1.114.105 --host 10.1.114.105 --client 242.1.0.0/16 --ikeport 4500 --dpdaction=hold --dpddelay 30]
002 "submariner-cable-dnd-prsurve-dev-2-10-1-114-105-0-0": added IKEv2 connection
181 "submariner-cable-dnd-prsurve-dev-2-10-1-114-105-0-0" #1: initiating IKEv2 connection
[90m2023-10-17T10:23:10.215Z[0m [32mINF[0m ..gine/cableengine.go:202 CableEngine          Successfully ins`

And this is the other cluster.

`[90m2023-10-17T13:00:53.875Z[0m [32mINF[0m ..-source/app/main.go:411 main                 Gateway leader election config values: main.leaderConfig{LeaseDuration:10, RenewDeadline:5, RetryPeriod:2}
[90m2023-10-17T13:00:53.875Z[0m [32mINF[0m ..-source/app/main.go:429 main                 Using namespace "submariner-operator" for the leader election lock
I1017 13:00:53.875555       1 leaderelection.go:245] attempting to acquire leader lease submariner-operator/submariner-gateway-lock...
[90m2023-10-17T13:00:55.755Z[0m [33mDBG[0m ..ry/request_handle.go:52 NAT                  Received request from 10.1.114.39:4490 - REQUEST_NUMBER: 0xd9a9855a8b5d99d, SENDER: "submariner-cable-dnd-prsurve-dev-1-10-1-114-39", RECEIVER: "submariner-cable-dnd-prsurve-dev-2-10-1-114-75"
[90m2023-10-17T13:00:55.755Z[0m [33mDBG[0m ..y/request_handle.go:117 NAT                  Sending response to 10.1.114.39:4490 - REQUEST_NUMBER: 0xd9a9855a8b5d99d, RESPONSE: OK, SENDER: "submariner-cable-dnd-prsurve-dev-2-10-1-114-75", RECEIVER: "submariner-cable-dnd-prsurve-dev-1-10-1-114-39"
I1017 13:01:32.027737       1 leaderelection.go:255] successfully acquired lease submariner-operator/submariner-gateway-lock
[90m2023-10-17T13:01:32.027Z[0m [33mDBG[0m ..ols/record/event.go:298 main                 Event(v1.ObjectReference{Kind:"Lease", Namespace:"submariner-operator", Name:"submariner-gateway-lock", UID:"5f585cbc-cb75-40a8-9e55-f5b2cc30a62b", APIVersion:"coordination.k8s.io/v1", ResourceVersion:"5819855", FieldPath:""}): type: 'Normal' reason: 'LeaderElection' compute-2-submariner-gateway became leader
[90m2023-10-17T13:01:32.028Z[0m [32mINF[0m ..reswan/libreswan.go:127 libreswan            Using NATT UDP port 4500
[90m2023-10-17T13:01:32.030Z[0m [32mINF[0m ..gine/cableengine.go:104 CableEngine          CableEngine controller started, driver: "libreswan"
[90m2023-10-17T13:01:32.030Z[0m [32mINF[0m ..public_ip_watcher.go:62 Endpoint             Starting the public IP watcher.
[90m2023-10-17T13:01:32.030Z[0m [32mINF[0m ..ers/tunnel/tunnel.go:40 Tunnel               Starting the tunnel controller
[90m2023-10-17T13:01:32.031Z[0m [32mINF[0m ..r/datastoresyncer.go:73 DSSyncer             Starting the datastore syncer
[90m2023-10-17T13:01:32.132Z[0m [32mINF[0m ..er/healthchecker.go:110 HealthChecker        CableEngine HealthChecker started with PingInterval: 1, MaxPacketLossCount: 5
[90m2023-10-17T13:01:32.134Z[0m [32mINF[0m ..thchecker/pinger.go:106 HealthChecker        Starting pinger for IP "242.0.255.254"
[90m2023-10-17T13:01:32.135Z[0m [32mINF[0m ..er/healthchecker.go:159 HealthChecker        CableEngine HealthChecker started pinger for CableName: "submariner-cable-dnd-prsurve-dev-1-10-1-114-39" with HealthCheckIP "242.0.255.254"
[90m2023-10-17T13:01:32.165Z[0m [32mINF[0m ..ery/natdiscovery.go:162 NAT                  Starting NAT discovery for endpoint "submariner-cable-dnd-prsurve-dev-1-10-1-114-39"
[90m2023-10-17T13:01:32.365Z[0m [32mINF[0m .._update_federator.go:72 Federator            local -> broker: Updated Cluster "submariner-broker/dnd-prsurve-dev-2" 
[90m2023-10-17T13:01:32.419Z[0m [32mINF[0m ../datastoresyncer.go:241 DSSyncer             Ensuring we are the only endpoint active for this cluster
[90m2023-10-17T13:01:32.420Z[0m [32mINF[0m ../datastoresyncer.go:316 DSSyncer             Creating local submariner Cluster: types.SubmarinerCluster{ID:"dnd-prsurve-dev-2", Spec:v1.ClusterSpec{ClusterID:"dnd-prsurve-dev-2", ColorCodes:[]string{"blue"}, ServiceCIDR:[]string{"172.30.0.0/16"}, ClusterCIDR:[]string{"10.128.0.0/14"}, GlobalCIDR:[]string{"242.1.0.0/16"}}} 
[90m2023-10-17T13:01:32.431Z[0m [32mINF[0m .._update_federator.go:72 Federator            broker -> local: Updated Cluster "submariner-operator/dnd-prsurve-dev-2" 
[90m2023-10-17T13:01:32.432Z[0m [32mINF[0m ../datastoresyncer.go:329 DSSyncer             Creating local submariner Endpoint: types.SubmarinerEndpoint{Spec:v1.EndpointSpec{ClusterID:"dnd-prsurve-dev-2", CableName:"submariner-cable-dnd-prsurve-dev-2-10-1-114-75", HealthCheckIP:"", Hostname:"compute-2", Subnets:[]string{"242.1.0.0/16"}, PrivateIP:"10.1.114.75", PublicIP:"66.187.232.130", NATEnabled:false, Backend:"libreswan", BackendConfig:map[string]string{"natt-discovery-port":"4490", "preferred-server":"false", "udp-port":"4500"}}} 
[90m2023-10-17T13:01:32.446Z[0m [32mINF[0m .._update_federator.go:72 Federator            broker -> local: Updated Endpoint "submariner-operator/dnd-prsurve-dev-2-submariner-cable-dnd-prsurve-dev-2-10-1-114-75" 
[90m2023-10-17T13:01:32.461Z[0m [32mINF[0m .._update_federator.go:72 Federator            local -> broker: Updated Endpoint "submariner-broker/dnd-prsurve-dev-2-submariner-cable-dnd-prsurve-dev-2-10-1-114-75" 
[90m2023-10-17T13:01:32.511Z[0m [32mINF[0m .._update_federator.go:72 Federator            local -> broker: Updated Endpoint "submariner-broker/dnd-prsurve-dev-2-submariner-cable-dnd-prsurve-dev-2-10-1-114-75" 
[90m2023-10-17T13:01:32.576Z[0m [32mINF[0m ../datastoresyncer.go:105 DSSyncer             Datastore syncer started
[90m2023-10-17T13:01:32.587Z[0m [32mINF[0m ..ncer/node_handler.go:73 DSSyncer             Updating the endpoint HealthCheckIP to globalIP "242.1.255.254"
[90m2023-10-17T13:01:32.588Z[0m [32mINF[0m ../datastoresyncer.go:329 DSSyncer             Creating local submariner Endpoint: types.SubmarinerEndpoint{Spec:v1.EndpointSpec{ClusterID:"dnd-prsurve-dev-2", CableName:"submariner-cable-dnd-prsurve-dev-2-10-1-114-75", HealthCheckIP:"242.1.255.254", Hostname:"compute-2", Subnets:[]string{"242.1.0.0/16"}, PrivateIP:"10.1.114.75", PublicIP:"66.187.232.130", NATEnabled:false, Backend:"libreswan", BackendConfig:map[string]string{"natt-discovery-port":"4490", "preferred-server":"false", "udp-port":"4500"}}} 
[90m2023-10-17T13:01:32.627Z[0m [32mINF[0m .._update_federator.go:72 Federator            broker -> local: Updated Endpoint "submariner-operator/dnd-prsurve-dev-2-submariner-cable-dnd-prsurve-dev-2-10-1-114-75" 
[90m2023-10-17T13:01:32.680Z[0m [32mINF[0m .._update_federator.go:72 Federator            local -> broker: Updated Endpoint "submariner-broker/dnd-prsurve-dev-2-submariner-cable-dnd-prsurve-dev-2-10-1-114-75" 
[90m2023-10-17T13:01:32.777Z[0m [33mDBG[0m ..ry/request_handle.go:52 NAT                  Received request from 10.1.114.39:4490 - REQUEST_NUMBER: 0xd9a9855a8b5d99f, SENDER: "submariner-cable-dnd-prsurve-dev-1-10-1-114-39", RECEIVER: "submariner-cable-dnd-prsurve-dev-2-10-1-114-75"
[90m2023-10-17T13:01:32.777Z[0m [33mDBG[0m ..y/request_handle.go:117 NAT                  Sending response to 10.1.114.39:4490 - REQUEST_NUMBER: 0xd9a9855a8b5d99f, RESPONSE: OK, SENDER: "submariner-cable-dnd-prsurve-dev-2-10-1-114-75", RECEIVER: "submariner-cable-dnd-prsurve-dev-1-10-1-114-39"
[90m2023-10-17T13:01:32.869Z[0m [33mDBG[0m ..ery/request_send.go:115 NAT                  Sending request - REQUEST_NUMBER: 0x13a6e38575ce4e52, SENDER: "submariner-cable-dnd-prsurve-dev-2-10-1-114-75", RECEIVER: "submariner-cable-dnd-prsurve-dev-1-10-1-114-39", USING_SRC: 10.1.114.75:4490, USING_DST: 10.1.114.39:4490
[90m2023-10-17T13:01:32.869Z[0m [33mDBG[0m ..ery/request_send.go:115 NAT                  Sending request - REQUEST_NUMBER: 0x13a6e38575ce4e53, SENDER: "submariner-cable-dnd-prsurve-dev-2-10-1-114-75", RECEIVER: "submariner-cable-dnd-prsurve-dev-1-10-1-114-39", USING_SRC: 10.1.114.75:4490, USING_DST: 66.187.232.130:4490
[90m2023-10-17T13:01:32.894Z[0m [33mDBG[0m ..y/response_handle.go:31 NAT                  Received response from 10.1.114.39:4490 - REQUEST_NUMBER: 0x13a6e38575ce4e52, RESPONSE: OK, SENDER: "submariner-cable-dnd-prsurve-dev-1-10-1-114-39", RECEIVER: "submariner-cable-dnd-prsurve-dev-2-10-1-114-75"
[90m2023-10-17T13:01:32.895Z[0m [33mDBG[0m ../remote_endpoint.go:184 NAT                  selected private IP "10.1.114.39" for endpoint "submariner-cable-dnd-prsurve-dev-1-10-1-114-39"
[90m2023-10-17T13:01:32.895Z[0m [32mINF[0m ..gine/cableengine.go:195 CableEngine          Installing Endpoint cable "submariner-cable-dnd-prsurve-dev-1-10-1-114-39"
[90m2023-10-17T13:01:32.895Z[0m [32mINF[0m ..reswan/libreswan.go:579 libreswan            Starting Pluto
002 listening for IKE messages
002 Kernel supports NIC esp-hw-offload
002 adding UDP interface tun0 10.131.0.1:500
002 adding UDP interface tun0 10.131.0.1:4500
002 adding UDP interface ens192 10.1.114.75:500
002 adding UDP interface ens192 10.1.114.75:4500
002 adding UDP interface lo 127.0.0.1:500
002 adding UDP interface lo 127.0.0.1:4500
002 adding UDP interface lo [::1]:500
002 adding UDP interface lo [::1]:4500
002 loading secrets from "/etc/ipsec.secrets"
002 loading secrets from "/etc/ipsec.d/submariner.secrets"
[90m2023-10-17T13:01:33.426Z[0m [32mINF[0m ..reswan/libreswan.go:371 libreswan            Creating connection(s) for {"metadata":{"name":"dnd-prsurve-dev-1-submariner-cable-dnd-prsurve-dev-1-10-1-114-39","namespace":"submariner-operator","uid":"c7551a0d-e0e8-4fae-af5c-1a47f29fe5e4","resourceVersion":"5579847","generation":6,"creationTimestamp":"2023-10-16T13:40:38Z","labels":{"submariner-io/clusterID":"dnd-prsurve-dev-1"},"managedFields":[{"manager":"submariner-gateway","operation":"Update","apiVersion":"submariner.io/v1","time":"2023-10-17T10:23:45Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:labels":{".":{},"f:submariner-io/clusterID":{}}},"f:spec":{".":{},"f:backend":{},"f:backend_config":{".":{},"f:natt-discovery-port":{},"f:preferred-server":{},"f:udp-port":{}},"f:cable_name":{},"f:cluster_id":{},"f:healthCheckIP":{},"f:hostname":{},"f:nat_enabled":{},"f:private_ip":{},"f:public_ip":{},"f:subnets":{}}}}]},"spec":{"cluster_id":"dnd-prsurve-dev-1","cable_name":"submariner-cable-dnd-prsurve-dev-1-10-1-114-39","healthCheckIP":"242.0.255.254","hostname":"compute-1","subnets":["242.0.0.0/16"],"private_ip":"10.1.114.39","public_ip":"66.187.232.130","nat_enabled":false,"backend":"libreswan","backend_config":{"natt-discovery-port":"4490","preferred-server":"false","udp-port":"4500"}}} in bi-directional mode
[90m2023-10-17T13:01:33.426Z[0m [32mINF[0m ..reswan/libreswan.go:435 libreswan            Executing whack with args: [--psk --encrypt --name submariner-cable-dnd-prsurve-dev-1-10-1-114-39-0-0 --id 10.1.114.75 --host 10.1.114.75 --client 242.1.0.0/16 --ikeport 4500 --to --id 10.1.114.39 --host 10.1.114.39 --client 242.0.0.0/16 --ikeport 4500 --dpdaction=hold --dpddelay 30]
002 "submariner-cable-dnd-prsurve-dev-1-10-1-114-39-0-0": added IKEv2 connection
181 "submariner-cable-dnd-prsurve-dev-1-10-1-114-39-0-0" #1: initiating IKEv2 connection
[90m2023-10-17T13:01:33.488Z[0m [32mINF[0m ..gine/cableengine.go:202 CableEngine          Successfully installed Endpoint cable "submariner-cable-dnd-prsurve-dev-1-10-1-114-39" with remote IP 10.1.114.39
[90m2023-10-17T13:01:33.527Z[0m [33mDBG[0m ..reswan/libreswan.go:255 libreswan            Connection "submariner-cable-dnd-prsurve-dev-1-10-1-114-39-0-0" not found in active connections obtained from whack: map[], map[]
[90m2023-10-17T13:01:33.527Z[0m [33mDBG[0m ..reswan/libreswan.go:269 libreswan            Connection "submariner-cable-dnd-prsurve-dev-1-10-1-114-39" not found in active connections obtained from whack: map[], map[]
[90m2023-10-17T13:01:34.997Z[0m [32mINF[0m .._update_federator.go:72 Federator            broker -> local: Updated Endpoint "submariner-operator/dnd-prsurve-dev-1-submariner-cable-dnd-prsurve-dev-1-10-1-114-39" 
[90m2023-10-17T13:01:35.008Z[0m [32mINF[0m .._update_federator.go:72 Federator            broker -> local: Updated Cluster "submariner-operator/dnd-prsurve-dev-1" 
[90m2023-10-17T13:01:38.135Z[0m [1m[31mERR[0m[0m ..thchecker/pinger.go:161 HealthChecker        Failed to successfully ping the remote endpoint IP "242.0.255.254" [36merror=[0m[31m"more than 5 packets lost"[0m
[90m2023-10-17T13:01:38.173Z[0m [32mINF[0m ..thchecker/pinger.go:177 HealthChecker        Ping to remote endpoint IP "242.0.255.254" is successful`

tpantelis commented:
I see where cluster "dnd-prsurve-dev-1" acquired the leader lock

` I1017 10:23:08.731886       1 leaderelection.go:255] successfully acquired lease submariner-operator/submariner-gateway-lock`

But it didn't lose the lock, at least not in that log snippet. I don't see anything out of the ordinary. How many gateways were configured on cluster "dnd-prsurve-dev-1"?

tpantelis commented:
This is related to https://issues.redhat.com/browse/ACM-8274. There were 3 gateways on cluster 2. `gateway-f6pm7 `got the leader lock first, created EP 10.1.114.75 and then lost the lock:

```
I1017 13:00:12.587547    1 leaderelection.go:255] successfully acquired lease submariner-operator/submariner-gateway-lock
E1017 13:00:26.130636    1 leaderelection.go:327] error retrieving resource lock submariner-operator/submariner-gateway-lock: Get “[https://172.30.0.1:443/apis/coordination.k8s.io/v1/namespaces/submariner-operator/leases/submariner-gateway-lock](https://172.30.0.1/apis/coordination.k8s.io/v1/namespaces/submariner-operator/leases/submariner-gateway-lock)”: context deadline exceeded
```

Then gateway-8647c got the leader lock and deleted EP 10.1.114.75 and created EP 10.1.114.99:

```
I1017 13:00:30.961641    1 leaderelection.go:255] successfully acquired lease submariner-operator/submariner-gateway-lock
[90m2023-10-17T13:00:31.292Z[0m [32mINF[0m ../datastoresyncer.go:265 DSSyncer             Successfully deleted existing submariner Endpoint “dnd-prsurve-dev-2-submariner-cable-dnd-prsurve-dev-2-10-1-114-75”
[90m2023-10-17T13:00:31.311Z[0m [32mINF[0m ../datastoresyncer.go:329 DSSyncer             Creating local submariner Endpoint: types.SubmarinerEndpoint{Spec:v1.EndpointSpec{ClusterID:“dnd-prsurve-dev-2", CableName:“submariner-cable-dnd-prsurve-dev-2-10-1-114-99”, HealthCheckIP:“”, Hostname:“compute-1”, Subnets:[]string{“242.1.0.0/16"}, PrivateIP:“10.1.114.99”, PublicIP:“66.187.232.130", NATEnabled:false, Backend:“libreswan”, BackendConfig:map[string]string{“natt-discovery-port”:“4490”, “preferred-server”:“false”, “udp-port”:“4500”}}}

```

But then `gateway-f6pm7` recreated EP 10.1.114.75 after it lost leadership and before it exited:

```
[90m2023-10-17T13:00:32.610Z[0m [32mINF[0m ../datastoresyncer.go:329 DSSyncer             Creating local submariner Endpoint: types.SubmarinerEndpoint{Spec:v1.EndpointSpec{ClusterID:“dnd-prsurve-dev-2”, CableName:“submariner-cable-dnd-prsurve-dev-2-10-1-114-75", HealthCheckIP:“242.1.255.254”, Hostname:“compute-2", Subnets:[]string{“242.1.0.0/16”}, PrivateIP:“10.1.114.75", PublicIP:“66.187.232.130”, NATEnabled:false, Backend:“libreswan”, BackendConfig:map[string]string{“natt-discovery-port”:“4490", “preferred-server”:“false”, “udp-port”:“4500"}}}
[9:38](https://redhat-internal.slack.com/archives/D052PTTFS0P/p1698370689903849)
[90m2023-10-17T13:00:33.054Z[0m [1m[31mFTL[0m[0m ..al/pkg/log/logger.go:67 main                 Leader election lost, shutting down
```

That’s odd. After leader election was lost it should’ve exited quickly but did not for 7 sec. And it should not have recreated the EP.

This occurred when updating the HealthCheckIP  in the EP:

`[90m2023-10-17T13:00:32.610Z[0m [32mINF[0m ..ncer/node_handler.go:73 DSSyncer             Updating the endpoint HealthCheckIP to globalIP “242.1.255.254”`

It does a create or update operation but we just want update in this case.

tpantelis commented:
I'll push a PR to fix the EP create/update issue.

sridhargaddam commented:
>That’s odd. After leader election was lost it should’ve exited quickly but did not for 7 sec. And it should not have recreated the EP.

Right, this is exactly what was happening. 

> It does a create or update operation but we just want update in this case.

AFAIR, we were getting update event itself for the endpoint?

tpantelis commented:
> AFAIR, we were getting update event itself for the endpoint?

Not sure I follow...

aswinsuryan commented:
> > That’s odd. After leader election was lost it should’ve exited quickly but did not for 7 sec. And it should not have recreated the EP.
> 
> Right, this is exactly what was happening.
> 
> > It does a create or update operation but we just want update in this case.
> 
> AFAIR, we were getting update event itself for the endpoint?

I think since we call create or update, when the new gateway node comes up and try to create the endpoint , it comes as an update event, since the endpoint already exists.

sridhargaddam commented:
> > AFAIR, we were getting update event itself for the endpoint?
> 
> Not sure I follow...

Saw your PR, its clear now.
